### DISCUSSION TITLE
Enhancement: Support for deploying librechat on a /subdirectory

### DISCUSSION THREAD

**Discussion by jordantgh** [danny-avila/LibreChat#1701] [Feature Requests & Suggestions]
### What features would you like to see added?

Currently, as far as I can tell, it is not possible to host librechat on a website as part of a subdirectory, e.g., "www.mysite.com/librechat" is not possible, as the code for getting assets seems to depend on absolute paths.

### More details

Minimal anonymised version of my nginx config.
    
```
server {
    server_name mysite.com www.mysite.com;

    location / {
       root /home/home/Public/homepage;
       index index.html;
   }

    location /librechat/api {
        proxy_pass http://localhost:3080/api/;
    }

    location /librechat {
        proxy_pass http://localhost:3080/;
    }

    listen 443 ssl;
    ssl_certificate /etc/letsencrypt/live/mysite.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/mysite.com/privkey.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf;
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;

}

server {
    listen 80;
    server_name mysite.com www.mysite.com;

    if ($host = www.mysite.com) {
        return 301 https://$host$request_uri;
    }

    if ($host = mysite.com) {
        return 301 https://$host$request_uri;
    }

    location ~ /.well-known/acme-challenge/ {
        allow all;
        root /var/www/html;
        try_files $uri =404;
        break;
    }

    return 404;
}
```

### Pictures

From behind reverse proxy:
![website subdirectory](https://github.com/danny-avila/LibreChat/assets/56029559/b2d4d3bb-0d1c-47c8-88c6-a708dc324281)

On localhost:
![localhost](https://github.com/danny-avila/LibreChat/assets/56029559/9edf9591-0bcc-4230-aa6d-f2ab54c6abd1)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
Thanks for your suggestion. I think this is a niche case but would be nice to have. I don't think it would be too involved to implement but it would need tinkering with a setup like yours. 

Given other issues, I'm not sure when I will get to it, but this would need some configuration as suggested here: https://stackoverflow.com/questions/77666915/how-to-run-a-vite-react-app-from-a-subfolder

PRs welcome for this

**Reply by jordantgh** [reply]
Thanks for the tip. I took a bash and failed, but I don't really know what I'm doing. On the off chance it's useful to anyone else, simply setting `basename` in `LibreChat/client/src/routes/index.tsx` and `base` in `LibreChat/client/vite.config.ts` is apparently insufficient (doing so got `localhost:3080/subfolder/` working, but not `mysite.com/subfolder` - index.html is still looking for assets at `/`).

**Reply by danny-avila** [reply]
> doing so got `localhost:3080/subfolder/` working

Nice!

Try changing this:
https://github.com/danny-avila/LibreChat/blob/f1d974c513b297e87b16a807d56a3cd7aafb52a6/api/server/index.js#L40-L41

to: 
```js
 app.use(express.static(path.join('your-basename', app.locals.paths.dist))); 
 app.use(express.static(path.join('your-basename', app.locals.paths.publicPath)));
```

That might get it working! Images pointing to the right filepath might be another issue to tackle but this would maybe get the bundled assets going.

**Reply by danny-avila** [reply]
or some combination of this that points to the right location. may need to prepend `your-basename` with `/` like this `/your-basename`

**Reply by jordantgh** [reply]
Sadly that didn't do it (I tried with/without the `/` prepended), it actually breaks localhost serving also. Sorry I don't have any ideas of my own.

**Reply by danny-avila** [reply]
gotcha, all I can suggest at the moment is searching the web, but you may need to edit the same file I pointed to and/or your nginx config


nginx with regex approach: https://stackoverflow.com/questions/22027622/serve-express-js-app-from-subfolder
express.js static serve edit: https://stackoverflow.com/questions/34597072/express-static-not-working-for-subdirectory

**Reply by thomsna** [reply]
@jordantgh Did you get this to work?

**Reply by jordantgh** [reply]
> @jordantgh Did you get this to work?

Nah I just decided to use a subdomain, which did work.

**Comment by jpsfs** [comment]
Just opened a PR implementing this feature: #9155 
Hope it gets merged soon 🙏

**Comment by alephreish** [comment]
Just started using LibreChat. @jpsfs Thanks a lot for this feature! I use Nginx as a reverse proxy and most of the stuff seems to work flawlessly with:

```
# .env
DOMAIN_CLIENT=http://localhost:3080/librechat
DOMAIN_SERVER=http://localhost:3080/librechat

# /etc/nginx/sites-available/default
location /librechat/ {
        proxy_pass http://localhost:3080/;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
}
```

The only thing is that the `filepaths` returned by `https://my_host/librechat/api/files` all come out without the subfolder `/librechat/` which results in dead links.

Is something missing on my side?

**Reply by CompartPKU** [reply]
I assume this is related to the problem that we are experiencing: We are running LibreChat on a subdirectory. When we upload a custom icon for an agent, it will not be displayed afterwards, because the icon file path is missing the subfolder part.

Our nginx configuration is analogous to the above.
